components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java

```diff
@@ -232,4 +250,45 @@ public OAuthApplicationInfo createOAuthApplication(
         return null;
     }
 
+    public void unregisterApplication(String userId, String applicationName, String consumerKey)
+            throws APIManagementException {
+
+        String tenantDomain = MultitenantUtils.getTenantDomain(userId);
+        String baseUser = CarbonContext.getThreadLocalCarbonContext().getUsername();
+        String userName = MultitenantUtils.getTenantAwareUsername(userId);
+
+        PrivilegedCarbonContext.startTenantFlow();
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain, true);
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setUsername(userName);
+
+        if (userId == null || userId.isEmpty()) {
+            throw new APIManagementException("Error occurred while unregistering Application: userId cannot be null/empty");
+        }
+        try {
+            OAuthAdminService oAuthAdminService = new OAuthAdminService();
+            OAuthConsumerAppDTO oAuthConsumerAppDTO = oAuthAdminService.getOAuthApplicationData(consumerKey);
+
+            if (oAuthConsumerAppDTO == null) {
+                throw new APIManagementException("Couldn't retrieve OAuth Consumer Application associated with the " +
+                                                 "given consumer key: " + consumerKey);
+            }
+            oAuthAdminService.removeOAuthApplicationData(consumerKey);
+
+            ApplicationManagementService appMgtService = ApplicationManagementService.getInstance();
+            ServiceProvider createdServiceProvider = appMgtService.getApplication(applicationName);
+
+            if (createdServiceProvider == null) {
+                throw new APIManagementException("Couldn't retrieve Service Provider Application " + applicationName);
+            }
+            appMgtService.deleteApplication(applicationName);
+
+        } catch (IdentityApplicationManagementException e) {
+            APIUtil.handleException("Error occurred while removing ServiceProvider for app " + applicationName, e);
+        } catch (Exception e) {
+            APIUtil.handleException("Error occurred while removing OAuthApp " + applicationName, e);
+        } finally {
+            PrivilegedCarbonContext.endTenantFlow();
+            PrivilegedCarbonContext.getThreadLocalCarbonContext().setUsername(baseUser);
```

Comment:
When we registering Oauth application, we create them under specific tenant for a given application owner. This owner is only known by this tenant. Therefor before we do anything specific to this tenant, we keep the current thread local user as a base user. Then we start tenant flow and set the application owner as the thread local user in order to use for the tenant specific logic. Finally, after ending the tenant flow, we set back the base user as the thread local user.